Tukio 4 per topic event broker

lhavel opened the pull request:
This work mainly adds the ability to pass a tuple of `(data, topic)` to the workflow engine.
Previously, the workflow engine tried to trigger a workflow each time it received a new data set (we call it an _event_) and it passed this data to all callbacks registered in the internal event broker.
From now on, it is possible to make an association between a data set and a _topic_ (something like a _channel_ in a pub/sub):
* Trigger workflows only when data is available in user-defined topics (can also be all data received or none)
* Register per-topic or global callbacks so that running tasks can get new data sets at runtime while avoiding to implement complex filtering in the implementation of the task itself.

The configuration dictionaries of a workflow has been slightly updated to add a new `topics` parameter at the workflow level and at the task level.

lerignoux commented on tukio/broker.py:
```diff
@@ -42,36 +48,115 @@ class Broker(object):
+            try:
+                self._handlers[coro_or_cb].add(topic)
+            except KeyError:
+                self._handlers[coro_or_cb] = {topic}
```
we may have concurrency problems

lhavel replied:
Do you mean in the context of a multi-threaded app?
This is designed to be run by a single event loop.

lerignoux commented on tukio/broker.py:
```diff
@@ -42,36 +48,115 @@ class Broker(object):
         self._loop = loop or asyncio.get_event_loop()
-        self._handlers = []
+        self._handlers = dict()
+        self._topics = {None: set()}
```
I don't really like having to maintain , update and check both.
Have you considered only the topics dictionary using a property ._handlers with the inverted dictionary ? 
return {v: k for k, v in self._topics.items()}

lhavel replied:
I found having both associations into two dict more clear to the dev. And it does not make any difference from the logic point of view nor from the performance point of view.
Besides, `_handlers` is not the exact reverse of `_topics`...

lerignoux commented on tukio/broker.py:
```diff
@@ -42,36 +48,115 @@ class Broker(object):
+        # Handler already registered with topics?
+        curr_topics = self._handlers.get(coro_or_cb)
+        if topic is not None:
+            if curr_topics == set():
```
why not None in curr_topics instead of an empty set() ?

lhavel replied:
This is a proposition of implementation explained in several docstrings:
* None means "listen to all topics"
* Empty set means "listen to no nothing at all"
* Not empty set means "listen to the given list of topics"
We talked about that with @thavel but it may worth a new discussion.

lerignoux commented on tukio/broker.py:
```diff
@@ -42,36 +48,115 @@ class Broker(object):
-        except ValueError:
-            warn = '{} is not registered in broker, cannot unregister'
-            log.warning(warn.format(coro_or_cb))
+        ke_exc = None
```
ke_exc ? what does ke means ? to clarify

lhavel replied:
`ke_exc` stands for _KeyError exception_

lerignoux commented on tukio/broker.py:
```diff
@@ -42,36 +48,115 @@ class Broker(object):
+                try:
+                    self._discard_topic(coro_or_cb, ct)
+                except KeyError as exc:
+                    ke_exc = exc
```
IMO this should not be permitted, consistency should always be ensured

lhavel replied:
Consistency is ensured by the broker! But as we can almost never guarantee we're bug free, I decided to implement it like that so that a production host wouldn't be too affected by a potential bug (ensure there's no memory leak in particular).

lerignoux commented on tukio/engine.py:
```diff
@@ -5,24 +5,139 @@
+        self.new = new_tmpl
+        self.current = current_tmpl
+
+    def __str__(self):
```
could be useful to keep the exception message usage to be able to add additional content if needed (concatenate err and self.message)

lhavel replied:
yep, why not

lerignoux commented on tukio/engine.py:
```diff
@@ -5,24 +5,139 @@
+            self._templates[template.uid] = template
+            current = None
+        else:
+            if current.version >= template.version:
```
if current is None we have an exception here.

lerignoux commented on tukio/engine.py:
```diff
@@ -5,24 +5,139 @@
+            if current.version >= template.version:
+                raise LoadWorkflowError(template, current)
+            else:
+                self._templates[template.uid] = template
```
meaning we cannot have two versions of the same template running.

Couldn't that be a problem ?

lhavel replied:
Yep this is a decision we took some days ago. I didn't change that in this PR. It was already implemented like that before.
The requirement to support multiple versions of the same template loaded in the engine must still be analyzed.

lerignoux commented on tukio/engine.py:
```diff
@@ -5,24 +5,139 @@
+        # Update the template ID/topics association
+        if current is not None:
+            self.unload(current)
+        topics = template.topics
```
shouldn't it return always a list ?

lhavel replied:
See the discussion with @thavel about that. We can talk about it to decide what to do.

lerignoux commented on tukio/engine.py:
```diff
@@ -43,7 +161,7 @@ def _add_wflow(self, wflow):
             self._running[wflow.template_id] = [wflow]
         self._running_by_id[wflow.uid] = wflow
-        log.debug('new workflow added to the running list: {}'.format(wflow))
+        log.debug('new workflow started %s', wflow)
```
still don't like this arguments passing method. why not a format ?

lhavel replied:
Because this is the recommended way to do it by the Python community (and the logging lib is implemented to do so)
Your linter should complain about that!
By the way it's more compact that the `.format()` method.

lerignoux commented on tukio/task/template.py:
```diff
@@ -48,21 +49,37 @@ def from_dict(cls, task_dict):
+        The parameters 'topics' and 'config' are both optional.
+        See below the behavior of a task at runtime according to the value of
+        'topics':
+            {"topics": None}
```
as stated before, I'd rather [None] to have a consistent pattern and much simpler code.

lhavel replied:
again the same comment. Let's talk about that.

lhavel replied:
As discussed last Friday, I kept the `topics` configuration conventions but I've now improved the way it is implemented to make the behaviors of tasks and engine more clear in the code.

lerignoux commented on tukio/workflow.py:
```diff
@@ -344,14 +413,16 @@ def _new_task(self, task_tmpl, inputs):
             return None
         else:
-            log.debug('new task created for {}'.format(task_tmpl))
+            log.debug('new task created for %s', task_tmpl)
```
still prefet the format (cf dev guidelines !)

lhavel replied:
See my previous comment about that.

lhavel commented:
I added a `Listen` enum to convert the `topics` configuration parameter into a clear expected behavior of the tasks and the engine (workflow selection).
Besides, I also cleaned up the code to remove any reference to execution objects from `TaskTemplate`.

thavel commented on tukio/broker.py:
```diff
@@ -42,36 +49,74 @@ class Broker(object):
+                raise ValueError('{} already registered as global'
+                                 ' handler'.format(coro_or_cb))
+            try:
+                self._topic_handlers[topic].add(coro_or_cb)
```
You remove the `try...except` statement:
```json
topic_handlers = self._topic_handlers.get(topic, {}).add(coro_or_cb)
self._topic_handlers[topic] = topic_handlers
```
